File: tools/gui/src/domain/schemas/viewpoints.ts
import { Schema } from 'effect'

import { TraceTableSchema } from './viewpointTrace'

export type { PatternResult, TraceObligation, TraceRow, TraceTable } from './viewpointTrace'

// ── Viewpoints (GUI selector/overlay) ─────────────────────────────────────────

export const ViewpointApplicationSchema = Schema.Struct({
  slug: Schema.String,
  version: Schema.Number,
  enforcement_override: Schema.optional(Schema.Literal('off', 'warn', 'ghost')),
  derivation_params: Schema.optional(Schema.Record({ key: Schema.String, value: Schema.Unknown })),
})
export type ViewpointApplication = typeof ViewpointApplicationSchema.Type

export const ViewpointSummarySchema = Schema.Struct({
  slug: Schema.String,
  version: Schema.Number,
  name: Schema.String,
  description: Schema.String,
  purpose: Schema.Array(Schema.String),
  content: Schema.Array(Schema.String),
  scope: Schema.Unknown,
})
export type ViewpointSummary = typeof ViewpointSummarySchema.Type

/** A `mode: "scale"` style rule's per-item value: interpolate between `tokens` at
 * `position` (0..1) — never a discrete token, since the rule declares a continuous
 * spectrum rather than named bands. */
export const ScaleStyleValueSchema = Schema.Struct({
  position: Schema.Number,
  tokens: Schema.Tuple(Schema.String, Schema.String),
})
export type ScaleStyleValue = typeof ScaleStyleValueSchema.Type

export const StyleValueSchema = Schema.Union(Schema.String, ScaleStyleValueSchema)
export type StyleValue = typeof StyleValueSchema.Type

export const ProjectedOccurrenceSchema = Schema.Struct({
  item_id: Schema.String,
  item_kind: Schema.Literal('entity', 'connection'),
  state: Schema.Literal('visible', 'ghosted'),
  membership: Schema.Literal('primary', 'expanded'),
  reasons: Schema.Array(Schema.Literal('out_of_scope', 'criteria_mismatch', 'endpoint_excluded')),
  style: Schema.Record({ key: Schema.String, value: StyleValueSchema }),
})
export type ProjectedOccurrence = typeof ProjectedOccurrenceSchema.Type

export const ScaleLegendDataSchema = Schema.Struct({
  capability: Schema.String,
  attribute: Schema.String,
  minimum: Schema.Number,
  maximum: Schema.Number,
  tokens: Schema.Array(Schema.String),
})
export type ScaleLegendData = typeof ScaleLegendDataSchema.Type

/** One authored style rule's observable outcome for an execution — the "no silent
 * no-op" contract. `expected-empty` is a legitimate state rendered as a quiet badge;
 * `unresolvable` and `shadowed` also arrive as warnings. */
export const StyleRuleOutcomeSchema = Schema.Struct({
  rule_index: Schema.Number,
  capability: Schema.String,
  kind: Schema.Literal('applied', 'expected-empty', 'shadowed', 'unresolvable', 'disabled'),
  matched_count: Schema.Number,
  applied_count: Schema.Number,
  detail: Schema.NullOr(Schema.String),
})
export type StyleRuleOutcome = typeof StyleRuleOutcomeSchema.Type

export const ViewpointProjectionSchema = Schema.Struct({
  applied: Schema.Boolean,
  target: Schema.optional(Schema.Literal('repository', 'diagram', 'matrix')),
  items: Schema.optional(Schema.Array(ProjectedOccurrenceSchema)),
  stale_pin: Schema.optional(Schema.Boolean),
  warnings: Schema.optional(Schema.Array(Schema.String)),
  scale_legends: Schema.optional(Schema.Array(ScaleLegendDataSchema)),
  rule_outcomes: Schema.optional(Schema.Array(StyleRuleOutcomeSchema)),
})
export type ViewpointProjection = typeof ViewpointProjectionSchema.Type

// ── Viewpoint execution ────────────────────────────────────────────────────────
// Fixed, non-customizable per-item summaries — deliberately unstyled.
// Style tokens come from the separate `execute-projection` endpoint (`ViewpointProjectionSchema`
// above), never from this contract, so MCP/REST content can never disagree with the GUI.

export const EntityItemSummarySchema = Schema.Struct({
  id: Schema.String,
  name: Schema.String,
  type: Schema.String,
  specialization_slugs: Schema.Array(Schema.String),
  group: Schema.String,
  membership: Schema.Literal('primary', 'expanded'),
  status: Schema.optionalWith(Schema.String, { default: () => '' }),
  version: Schema.optionalWith(Schema.String, { default: () => '' }),
  /** Present when the definition authors table columns: one entry per column source,
   * resolved server-side at the execution's snapshot; a source with no value for this
   * entity is explicitly null. */
  column_values: Schema.optionalWith(
    Schema.NullOr(Schema.Record({ key: Schema.String, value: Schema.NullOr(Schema.Unknown) })),
    { default: () => null },
  ),
  /** Modeled hop distance from the nearest anchor (0 = anchor, 1 = direct modeled edge,
   * N = minimum derived witness-chain length). Null/absent when the execution is
   * unanchored or the entity has no connecting edge to any anchor (unranked — rendered
   * as its own state, never as distance 0 or 1). */
  anchor_modeled_distance: Schema.optionalWith(Schema.NullOr(Schema.Number), { default: () => null }),
  /** Set when this entity's criteria match required derived-relationship evidence: the
   * minimum witness-chain length the verdict rested on. Null when the match holds on
   * modeled facts alone — surfaces tag such rows "matched via derived (N hops)". */
  matched_via_derived_hops: Schema.optionalWith(Schema.NullOr(Schema.Number), { default: () => null }),
})
export type EntityItemSummary = typeof EntityItemSummarySchema.Type

/** One step of a derived connection's witness chain, already ordered source→target by
 * the server (`via_connection_ids` is unordered membership and never renderable as-is). */
export const WitnessStepSchema = Schema.Struct({
  connection_id: Schema.String,
  source: Schema.String,
  target: Schema.String,
  connection_type: Schema.String,
  direction: Schema.Literal('forward', 'reverse'),
  hop_index: Schema.Number,
})
export type WitnessStep = typeof WitnessStepSchema.Type

export const ConnectionItemSummarySchema = Schema.Struct({
  id: Schema.String,
  type: Schema.String,
  source: Schema.String,
  target: Schema.String,
  certainty: Schema.NullOr(Schema.Literal('certain', 'potential')),
  hops: Schema.NullOr(Schema.Number),
  via_connection_ids: Schema.Array(Schema.String),
  /** Empty for modeled connections, and for a derived connection whose chain could not
   * be reconstructed (renderers show that as "chain unavailable", never as no chain). */
  witness_steps: Schema.optionalWith(Schema.Array(WitnessStepSchema), { default: () => [] }),
})
export type ConnectionItemSummary = typeof ConnectionItemSummarySchema.Type

export const MatrixAxisIdsSchema = Schema.Struct({
  row_entity_ids: Schema.Array(Schema.String),
  column_entity_ids: Schema.Array(Schema.String),
})
export type MatrixAxisIds = typeof MatrixAxisIdsSchema.Type

/** Classification of the FULL result against the definition's declared target
 * population — absent when the target population is unknown, in which case headers show
 * plain counts and make NO absence claims. */
export const AggregateNodeSchema = Schema.Struct({
  id: Schema.String,
  dimension: Schema.Literal('group', 'domain', 'type'),
  dimension_value: Schema.String,
  entity_type: Schema.String,
  member_count: Schema.Number,
  member_ids: Schema.Array(Schema.String),
})
export type AggregateNode = typeof AggregateNodeSchema.Type

export const AggregateEdgeSchema = Schema.Struct({
  id: Schema.String,
  source_aggregate_id: Schema.String,
  target_aggregate_id: Schema.String,
  connection_type: Schema.String,
  provenance: Schema.Literal('modeled', 'derived-certain', 'derived-potential'),
  member_count: Schema.Number,
  member_connection_ids: Schema.Array(Schema.String),
})
export type AggregateEdge = typeof AggregateEdgeSchema.Type

/** Present when the complete population exceeds the effective legibility budget on a
 * graph surface: the super-nodes and bundled edges the surface opens with. */
export const AggregationSummarySchema = Schema.Struct({
  dimension: Schema.Literal('group', 'domain', 'type'),
  legibility_budget: Schema.Number,
  nodes: Schema.Array(AggregateNodeSchema),
  edges: Schema.Array(AggregateEdgeSchema),
})
export type AggregationSummary = typeof AggregationSummarySchema.Type

export const TargetPopulationSummarySchema = Schema.Struct({
  target_types: Schema.Array(Schema.String),
  target_count: Schema.Number,
  incidental_type_counts: Schema.Record({ key: Schema.String, value: Schema.Number }),
  structural_count: Schema.Number,
})
export type TargetPopulationSummary = typeof TargetPopulationSummarySchema.Type

export const ViewpointExecutionResultSchema = Schema.Struct({
  slug: Schema.NullOr(Schema.String),
  version: Schema.NullOr(Schema.Number),
  query_schema: Schema.Number,
  repo_scope: Schema.String,
  executed_at: Schema.String,
  index_generation: Schema.NullOr(Schema.Number),
  entity_ids: Schema.Array(Schema.String),
  connection_ids: Schema.Array(Schema.String),
  entities: Schema.Array(EntityItemSummarySchema),
  connections: Schema.Array(ConnectionItemSummarySchema),
  total_entity_count: Schema.Number,
  returned_entity_count: Schema.Number,
  total_connection_count: Schema.Number,
  returned_connection_count: Schema.Number,
  truncated: Schema.Boolean,
  entity_limit: Schema.Number,
  matrix_axes: Schema.NullOr(MatrixAxisIdsSchema),
  warnings: Schema.Array(Schema.String),
  duration_ms: Schema.Number,
  query_summary: Schema.String,
  /** Entity ids the execution was anchored on (resolved `entity-id` parameter values) —
   * presentations mark/center these and derive hop distances from them. */
  anchor_ids: Schema.optionalWith(Schema.Array(Schema.String), { default: () => [] }),
  target_population: Schema.optionalWith(Schema.NullOr(TargetPopulationSummarySchema), { default: () => null }),
  aggregation: Schema.optionalWith(Schema.NullOr(AggregationSummarySchema), { default: () => null }),
  /** The canonical values this execution ran with — what a shared URL must reproduce. */
  bound_parameters: Schema.optionalWith(Schema.Record({ key: Schema.String, value: Schema.Unknown }), {
    default: () => ({}),
  }),
  trace_table: Schema.optionalWith(Schema.NullOr(TraceTableSchema), { default: () => null }),
})
export type ViewpointExecutionResult = typeof ViewpointExecutionResultSchema.Type

export interface ViewpointExecutionRequest {
  readonly slug?: string
  readonly query?: unknown
  readonly limit?: number
  readonly parameters?: Record<string, unknown>
}

/** GUI-only ad-hoc ArchiMate-notation rendering (the ad-hoc `diagram` representation) —
 * unstyled; `node_color`/`edge_color`/`edge_emphasis` overlays are applied client-side
 * onto the returned SVG. */
export const SignalBannerSchema = Schema.Struct({
  classification: Schema.NullOr(Schema.String),
  available: Schema.Boolean,
  note: Schema.NullOr(Schema.String),
  basis_snapshots: Schema.Array(Schema.Struct({
    anchor_entity_id: Schema.String,
    snapshot_id: Schema.String,
    activated_at: Schema.String,
  })),
  generated_at: Schema.String,
})
export type SignalBanner = typeof SignalBannerSchema.Type

export const ViewpointDiagramResultSchema = Schema.Struct({
  svg: Schema.NullOr(Schema.String),
  warnings: Schema.Array(Schema.String),
  // Entity id -> the rendered SVG's PlantUML alias (never the raw artifact id) — needed to
  // resolve SVG elements back to artifact ids for click-to-select, the same way a real
  // persisted diagram's viewer already does from its own diagram_entities.
  entity_aliases: Schema.optional(Schema.Record({ key: Schema.String, value: Schema.String })),
  // Present ONLY for definitions declaring a security-signal source: computed
  // classification + basis snapshots + generation timestamp (the D11 ephemeral render).
  signal_banner: Schema.optional(Schema.NullOr(SignalBannerSchema)),
})
export type ViewpointDiagramResult = typeof ViewpointDiagramResultSchema.Type

// ── Viewpoint definitions (management view) ───────────────────────────────────
// `query`/`presentation`/`scope` stay `Unknown` here (matching the existing
// `ViewpointSummarySchema.scope` convention above) — their recursive wire shape is
// modeled by plain TS types in `viewpointCriteria.ts`, decoded/encoded by
// `viewpointCriteriaSerialization.ts`, not by Effect Schema.

export const ScopeSummarySchema = Schema.Struct({
  unrestricted: Schema.Boolean,
  entity_types: Schema.optional(Schema.Array(Schema.String)),
  connection_types: Schema.optional(Schema.Array(Schema.String)),
  excluded_entity_types: Schema.optional(Schema.Array(Schema.String)),
  excluded_domains: Schema.optional(Schema.Array(Schema.String)),
  excluded_connection_types: Schema.optional(Schema.Array(Schema.String)),
})
export type ScopeSummary = typeof ScopeSummarySchema.Type

export const ViewpointDefinitionEnvelopeSchema = Schema.Struct({
  slug: Schema.String,
  version: Schema.Number,
  name: Schema.String,
  description: Schema.optional(Schema.String),
  rationale: Schema.optional(Schema.String),
  purpose: Schema.optional(Schema.Union(Schema.String, Schema.Array(Schema.String))),
  content: Schema.optional(Schema.Union(Schema.String, Schema.Array(Schema.String))),
  stakeholders: Schema.optional(Schema.Array(Schema.String)),
  concerns: Schema.optional(Schema.Array(Schema.String)),
  scope: Schema.optional(Schema.Unknown),
  representation_types: Schema.optional(Schema.Array(Schema.String)),
  derivation_defaults: Schema.optional(Schema.Record({ key: Schema.String, value: Schema.Unknown })),
  query: Schema.optional(Schema.Unknown),
  presentation: Schema.optional(Schema.Unknown),
  /** Which selection layer is ACTIVE (scope | query); absent on pre-migration
   * definitions, where the legacy behavior (query when present, else scope) applies. */
  selection_mode: Schema.optional(Schema.Literal('scope', 'query')),
  /** Fork provenance, stamped server-side at fork time (origin slug/version/content
   * digest); absent on non-forks. */
  forked_from: Schema.optional(Schema.Struct({
    slug: Schema.String,
    version: Schema.Number,
    definition_digest: Schema.String,
    index_generation: Schema.optional(Schema.Number),
  })),
  /** Digest-computed staleness against the CURRENT origin — 'stale' the moment the
   * origin's content changes, even without a version bump. Null for non-forks. */
  fork_status: Schema.optionalWith(
    Schema.NullOr(Schema.Literal('current', 'stale', 'origin-missing')),
    { default: () => null },
  ),
  /** The definition's CURRENT canonical content digest — verified execution references
   * pin it so a later open can say the definition changed. */
  definition_digest: Schema.optional(Schema.String),
  tier: Schema.Literal('module', 'enterprise', 'engagement'),
  scope_summary: ScopeSummarySchema,
  query_summary: Schema.NullOr(Schema.String),
})
export type ViewpointDefinitionEnvelope = typeof ViewpointDefinitionEnvelopeSchema.Type

export const ViewpointDefinitionListSchema = Schema.Struct({
  viewpoints: Schema.Array(ViewpointDefinitionEnvelopeSchema),
})

const BindingCatalogSchema = Schema.Struct({
  select: Schema.Array(Schema.String),
  aggregate: Schema.Array(Schema.String),
  result_types: Schema.Array(Schema.String),
})

const ParameterCatalogSchema = Schema.Struct({ types: Schema.Array(Schema.String) })

const DerivedCatalogSchema = Schema.Struct({
  traversal: Schema.Array(Schema.String),
  certainty: Schema.Array(Schema.String),
  reduce: Schema.Array(Schema.String),
})

const ConnectionDerivationEntrySchema = Schema.Struct({ role: Schema.String, strength: Schema.NullOr(Schema.Number) })

export const CriteriaCatalogSchema = Schema.Struct({
  entity_types: Schema.Array(Schema.String),
  connection_types: Schema.Array(Schema.String),
  specialization_slugs: Schema.Array(Schema.String),
  entity_attribute_types: Schema.Record({ key: Schema.String, value: Schema.String }),
  connection_attribute_types: Schema.Record({ key: Schema.String, value: Schema.String }),
  // Enumerable value sets per attribute path (schema-declared `enum` attributes, plus the
  // enumerable reserved facets `domain`/`status`) — drives the criteria value picker's
  // switch from free text to a dropdown / multi-select. Optional-with-default so the editor
  // tolerates a backend that predates this field (falls back to free-text everywhere).
  entity_attribute_enums: Schema.optionalWith(
    Schema.Record({ key: Schema.String, value: Schema.Array(Schema.String) }),
    { default: () => ({}) },
  ),
  connection_attribute_enums: Schema.optionalWith(
    Schema.Record({ key: Schema.String, value: Schema.Array(Schema.String) }),
    { default: () => ({}) },
  ),
  symmetric_connection_types: Schema.Array(Schema.String),
  reserved_entity_paths: Schema.Array(Schema.String),
  reserved_connection_paths: Schema.Array(Schema.String),
  depth_cap: Schema.Number,
  // entity type slug -> owning domain (hierarchy[0]) — lets the scope picker group entity
  // types by domain and support "exclude this whole domain" bulk actions.
  entity_type_domains: Schema.Record({ key: Schema.String, value: Schema.String }),
  // Registries snapshot for the bindings/parameters/derived-attribute panels' own pickers —
  // same "one snapshot, every picker" convention as the entity/connection type lists above.
  bindings: BindingCatalogSchema,
  parameters: ParameterCatalogSchema,
  derived: DerivedCatalogSchema,
  connection_derivation: Schema.Record({ key: Schema.String, value: ConnectionDerivationEntrySchema }),
})
export type CriteriaCatalog = typeof CriteriaCatalogSchema.Type

export const ViewpointSummarizeResultSchema = Schema.Struct({ summary: Schema.String })

export const ViewpointValidationIssueSchema = Schema.Struct({
  severity: Schema.Literal('error', 'warning'),
  code: Schema.String,
  path: Schema.String,
  message: Schema.String,
  expected: Schema.NullOr(Schema.String),
  found: Schema.NullOr(Schema.String),
})
export type ViewpointValidationIssue = typeof ViewpointValidationIssueSchema.Type

export const ViewpointReferencerSchema = Schema.Struct({
  artifact_id: Schema.String,
  target_kind: Schema.Literal('diagram', 'matrix'),
})
export type ViewpointReferencer = typeof ViewpointReferencerSchema.Type

export const ViewpointReferencerListSchema = Schema.Struct({
  referencers: Schema.Array(ViewpointReferencerSchema),
})

/** `pruned` slugs were pinned but have since dropped out of the effective catalog (e.g. a
 * deleted engagement-repo definition) — reported so the GUI can surface that a pin was
 * silently dropped, rather than the pin list just quietly shrinking. Only `GET` reports
 * it; `PUT`'s response echoes just the saved `slugs`. */
export const ViewpointPinsSchema = Schema.Struct({
  slugs: Schema.Array(Schema.String),
  pruned: Schema.optional(Schema.Array(Schema.String)),
})
export type ViewpointPins = typeof ViewpointPinsSchema.Type

export const ViewpointPersistResultSchema = Schema.Struct({
  ok: Schema.Boolean,
  action: Schema.Literal('create', 'edit', 'delete'),
  slug: Schema.String,
  version: Schema.NullOr(Schema.Number),
  dry_run: Schema.Boolean,
  issues: Schema.Array(ViewpointValidationIssueSchema),
  referencers: Schema.Array(ViewpointReferencerSchema),
})
export type ViewpointPersistResult = typeof ViewpointPersistResultSchema.Type

File: src/infrastructure/gui/routers/viewpoints.py
"""Read-only REST endpoints for execution, projection, and diagram previews."""

from __future__ import annotations

from dataclasses import asdict
from pathlib import Path
from typing import Annotated

from fastapi import APIRouter, Body, Depends, HTTPException

from src.application.runtime_catalogs import RuntimeCatalogs
from src.application.viewpoints.artifact_projection import project_artifact_by_frontmatter
from src.application.viewpoints.derived_connection_records import derived_connection_record
from src.application.viewpoints.evaluate_viewpoint import (
    UnknownViewpointSlugError,
    ViewpointExecutionRequest,
    ViewpointExecutionTimeoutError,
    evaluate_viewpoint,
    project_viewpoint_repository,
)
from src.application.viewpoints.parameter_binding import ViewpointParameterError
from src.application.viewpoints.registry_snapshot import build_registry_snapshot
from src.config.viewpoints_settings import (
    viewpoints_derivation_max_hops,
    viewpoints_derivation_max_relationships,
    viewpoints_derivation_time_budget_seconds,
    viewpoints_execution_max_entities,
    viewpoints_execution_timeout_seconds,
)
from src.domain.relationship_reachability import DerivationLimitError, is_derived_connection_id
from src.domain.viewpoint_binding_evaluation import BindingCardinalityError
from src.domain.viewpoint_condition_validation import RegistrySnapshot
from src.domain.viewpoint_query_parsing import query_from_mapping
from src.domain.viewpoints import TargetKind
from src.infrastructure.gui.routers import state as s
from src.infrastructure.gui.routers._diagram_selection import resolve_diagram_selection
from src.infrastructure.gui.routers._viewpoint_freshness import fresh_viewpoints_runtime_catalogs_dependency

router = APIRouter()

# Fixed notation for unpersisted diagram previews. Styling overlays are applied by the
# client to the returned SVG, so this endpoint returns unstyled notation only.
_AD_HOC_DIAGRAM_TYPE = "archimate-layered"


def _parameter_error(exc: ViewpointParameterError) -> HTTPException:
    return HTTPException(400, {"code": exc.code, "path": f"parameters/{exc.parameter}", "message": str(exc)})


def _execution_error(code: str, message: str, *, path: str = "query") -> HTTPException:
    return HTTPException(400, {"code": code, "path": path, "message": message})


def _registry_snapshot(catalogs: RuntimeCatalogs, repo_roots: list[Path]) -> RegistrySnapshot:
    return build_registry_snapshot(
        catalogs,
        repo_roots,
        derivation_max_hops=viewpoints_derivation_max_hops(),
        derivation_max_relationships=viewpoints_derivation_max_relationships(),
        derivation_time_budget_seconds=viewpoints_derivation_time_budget_seconds(),
    )


def _definition_label_attribute(slug: str | None, catalogs: RuntimeCatalogs) -> str | None:
    """The saved definition's ``label_attribute`` display option, when executing by slug —
    an ad-hoc query has no persisted presentation to read one from."""
    if slug is None:
        return None
    definition = catalogs.viewpoints.get(slug)
    if definition is None or definition.presentation is None:
        return None
    value = definition.presentation.display_options.get("label_attribute")
    return value if isinstance(value, str) and value else None


@router.post("/api/viewpoints/execute")
def execute_viewpoint(
    slug: Annotated[str | None, Body()] = None,
    query: Annotated[dict[str, object] | None, Body()] = None,
    limit: Annotated[int | None, Body()] = None,
    parameters: Annotated[dict[str, object] | None, Body()] = None,
    catalogs: RuntimeCatalogs = Depends(fresh_viewpoints_runtime_catalogs_dependency),
) -> dict[str, object]:
    """Execute a viewpoint by ``slug`` (catalog definition) or ``query`` (ad-hoc, no
    presentation/styling/column parameters. Its response matches the MCP result."""
    if (slug is None) == (query is None):
        raise HTTPException(400, "exactly one of 'slug' or 'query' must be provided")

    parsed_query = query_from_mapping(query, label="query") if query is not None else None
    request = ViewpointExecutionRequest(slug=slug, query=parsed_query, limit=limit, parameters=parameters)
    repo = s.get_repo()
    registries = _registry_snapshot(catalogs, repo.repo_roots)
    max_entities = viewpoints_execution_max_entities()
    try:
        result = evaluate_viewpoint(
            request,
            catalog=catalogs.viewpoints,
            read_access=repo,
            registries=registries,
            index_generation=repo.read_model_version().generation,
            max_entities=max_entities,
            default_limit=max_entities,
            timeout_seconds=viewpoints_execution_timeout_seconds(),
        )
    except UnknownViewpointSlugError as exc:
        raise HTTPException(400, str(exc)) from exc
    except ViewpointExecutionTimeoutError as exc:
        raise HTTPException(504, {"code": "execution-timeout", "path": "query", "message": str(exc)}) from exc
    except ViewpointParameterError as exc:
        raise _parameter_error(exc) from exc
    except BindingCardinalityError as exc:
        raise _execution_error(exc.code, str(exc)) from exc
    except DerivationLimitError as exc:
        raise _execution_error("derivation-limit", str(exc)) from exc
    return asdict(result)


@router.post("/api/viewpoints/execute-projection")
def execute_viewpoint_projection(
    slug: Annotated[str | None, Body()] = None,
    query: Annotated[dict[str, object] | None, Body()] = None,
    parameters: Annotated[dict[str, object] | None, Body()] = None,
    catalogs: RuntimeCatalogs = Depends(fresh_viewpoints_runtime_catalogs_dependency),
) -> dict[str, object]:
    """Return GUI projection items with style tokens for the selected population."""
    if (slug is None) == (query is None):
        raise HTTPException(400, "exactly one of 'slug' or 'query' must be provided")
    parsed_query = query_from_mapping(query, label="query") if query is not None else None
    repo = s.get_repo()
    registries = _registry_snapshot(catalogs, repo.repo_roots)
    index_generation = repo.read_model_version().generation
    try:
        projection = project_viewpoint_repository(
            slug,
            parsed_query,
            catalog=catalogs.viewpoints,
            read_access=repo,
            registries=registries,
            parameters=parameters,
        )
    except UnknownViewpointSlugError as exc:
        raise HTTPException(400, str(exc)) from exc
    except ViewpointParameterError as exc:
        raise _parameter_error(exc) from exc
    except BindingCardinalityError as exc:
        raise _execution_error(exc.code, str(exc)) from exc
    except DerivationLimitError as exc:
        raise _execution_error("derivation-limit", str(exc)) from exc
    # Same provenance contract as /execute: consumers correlating an execution result
    # with its styled projection can verify both came from the same model snapshot.
    return {"applied": True, "index_generation": index_generation, **asdict(projection)}


@router.post("/api/viewpoints/execute-diagram")
def execute_viewpoint_diagram(
    slug: Annotated[str | None, Body()] = None,
    query: Annotated[dict[str, object] | None, Body()] = None,
    parameters: Annotated[dict[str, object] | None, Body()] = None,
    catalogs: RuntimeCatalogs = Depends(fresh_viewpoints_runtime_catalogs_dependency),
) -> dict[str, object]:
    """Render an unpersisted ArchiMate diagram for the evaluated population."""
    if (slug is None) == (query is None):
        raise HTTPException(400, "exactly one of 'slug' or 'query' must be provided")
    parsed_query = query_from_mapping(query, label="query") if query is not None else None
    repo = s.get_repo()
    repo_root = s.maybe_engagement_root()
    if repo_root is None:
        raise HTTPException(500, "Repository not initialized")
    registries = _registry_snapshot(catalogs, repo.repo_roots)
    max_entities = viewpoints_execution_max_entities()
    request = ViewpointExecutionRequest(slug=slug, query=parsed_query, limit=max_entities, parameters=parameters)
    try:
        result = evaluate_viewpoint(
            request,
            catalog=catalogs.viewpoints,
            read_access=repo,
            registries=registries,
            index_generation=repo.read_model_version().generation,
            max_entities=max_entities,
            default_limit=max_entities,
            timeout_seconds=viewpoints_execution_timeout_seconds(),
        )
    except UnknownViewpointSlugError as exc:
        raise HTTPException(400, str(exc)) from exc
    except ViewpointExecutionTimeoutError as exc:
        raise HTTPException(504, {"code": "execution-timeout", "path": "query", "message": str(exc)}) from exc
    except ViewpointParameterError as exc:
        raise _parameter_error(exc) from exc
    except BindingCardinalityError as exc:
        raise _execution_error(exc.code, str(exc)) from exc
    except DerivationLimitError as exc:
        raise _execution_error("derivation-limit", str(exc)) from exc

    from src.application.artifact_parsing import normalize_puml_alias
    from src.infrastructure.rendering.diagram_builder import generate_archimate_puml_body, render_puml_svg

    modeled_connection_ids = [cid for cid in result.connection_ids if not is_derived_connection_id(cid)]
    entities, connections, _, _ = resolve_diagram_selection(repo, list(result.entity_ids), modeled_connection_ids)
    derived_records = [
        derived_connection_record(summary) for summary in result.connections if summary.certainty is not None
    ]
    puml = generate_archimate_puml_body(
        result.slug or "viewpoint-preview",
        entities,
        [*connections, *derived_records],
        diagram_type=_AD_HOC_DIAGRAM_TYPE,
        repo_root=repo_root,
        label_attribute=_definition_label_attribute(slug, catalogs),
    )
    svg, render_warnings = render_puml_svg(puml, repo_root, _AD_HOC_DIAGRAM_TYPE)
    # The rendered SVG's node/edge ids are PlantUML aliases (`normalize_puml_alias`'d from
    # each entity's `display_alias`), never the raw artifact id — the client-side click-to-
    # select overlay needs this mapping to resolve SVG elements back to artifact ids, the
    # same way a real persisted diagram's viewer already does from its own diagram_entities.
    entity_aliases = {e.artifact_id: normalize_puml_alias(e.display_alias) for e in entities if e.display_alias}
    return {"svg": svg, "warnings": [*result.warnings, *render_warnings], "entity_aliases": entity_aliases}


@router.get("/api/diagrams/{artifact_id}/viewpoint-projection")
def get_diagram_viewpoint_projection(
    artifact_id: str,
    catalogs: RuntimeCatalogs = Depends(fresh_viewpoints_runtime_catalogs_dependency),
) -> dict[str, object]:
    """Return the optional saved viewpoint projection for one diagram or matrix."""
    repo = s.get_repo()
    diag_rec = repo.get_diagram(artifact_id)
    if diag_rec is None:
        raise HTTPException(404, f"Diagram not found: {artifact_id!r}")
    target_kind: TargetKind = "matrix" if diag_rec.diagram_type == "matrix" else "diagram"
    module = catalogs.diagram_types.find_diagram_type(diag_rec.diagram_type if target_kind == "diagram" else "matrix")
    if module is None:
        raise HTTPException(404, f"Diagram type not found: {diag_rec.diagram_type!r}")
    _, registry, _ = s.get_write_deps()
    registries = _registry_snapshot(catalogs, repo.repo_roots)
    projection = project_artifact_by_frontmatter(
        diag_rec.extra,
        target_kind=target_kind,
        target_id=diag_rec.artifact_id,
        catalog=catalogs.viewpoints,
        module=module,
        entity_type_infos=catalogs.ontology.all_entity_types(),
        default_enforcement=catalogs.viewpoint_enforcement,
        registry=registry,
        registries=registries,
    )
    if projection is None:
        return {"applied": False}
    return {"applied": True, **asdict(projection)}
